src/v/pandaproxy/kafka_client_cache.cc

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "pandaproxy/kafka_client_cache.h"
+
+namespace pandaproxy {
+
+kafka_client_cache::kafka_client_cache(
+  YAML::Node const& cfg,
+  size_t size,
+  std::vector<config::broker_authn_endpoint> kafka_api,
+  model::timestamp::type keep_alive)
+  : _config{cfg}
+  , _cache_size{size}
+  , _kafka_has_sasl{false}
+  , _keep_alive{keep_alive} {
+    // Is there a Kafka listener with SASL enabled?
+    auto ep_it = std::find_if(
+      kafka_api.begin(),
+      kafka_api.end(),
+      [](const config::broker_authn_endpoint& ep) {
+          auto method = ep.authn_method.value_or(
+            config::broker_authn_method::none);
+          return method == config::broker_authn_method::sasl;
+      });
+
+    _kafka_has_sasl = ep_it != kafka_api.end();
+}
+
+client_ptr kafka_client_cache::fetch(credential_t user) {
+    auto it = _user_client_map.find(user.name);
+
+    // User not found
+    if (it == _user_client_map.end()) {
+        throw std::out_of_range(fmt::format("User {} not found", user.name));
+    }
+
+    // Otherwise user found. Move it to the beginning
+    // of the frequency list. Splice will shift the items
+    // over.
+    _user_client_list.splice(
+      _user_client_list.begin(), _user_client_list, it->second);
+
+    // The client is at the second of the iterator.
+    return it->second->second;
+}
+
+client_ptr kafka_client_cache::make_client(
+  credential_t user, config::rest_authn_type authn_type) {
+    // Set the principal when there is a Kafka listener with SASL
+    // enabled and the incoming request is using HTTP Basic AuthN
+    if (_kafka_has_sasl && authn_type == config::rest_authn_type::http_basic) {
+        // Need to specify type or else bad any_cast runtime error
+        _config.sasl_mechanism.set_value(ss::sstring{"SCRAM-SHA-256"});
+        _config.scram_username.set_value(user.name);
+        _config.scram_password.set_value(user.pass);
```

Comment:
it might be cleaner to make a copy of the common _config member and then update that copy passing it into the timestamped client rather than updating the member in place for each different user and then making a copy.

Reply:
I'll put it in.